students/navdeep/session06/mailroom_part4.py

```diff
@@ -0,0 +1,149 @@
+            while new_donation < 1:
+                new_donation = float(input("Enter something greater than 0: "))
+            if new_donor.strip() in donor_list:
+                donor_list[new_donor].append(new_donation)
+            else:
+                donor_list[new_donor.strip()] = [new_donation]
+            printThankYou(new_donor, new_donation)
+
+def showDonorNames(donor_list):
+    """
+    will show the unique names of donors.  Will use comprehension to print
+    out the donor names
+    :param donor_list: the list of donors and their donations
+    """
+    print("Here are the list of donors: ")
+    [print(donor) for donor in donor_list.keys()]
+    #for donor in donor_list.keys():
+     #   print(donor)
+
+def printThankYou(newest_donor, newest_donation):
+    """
+    Prints out a thank you message after someone has donated
+    :param newest_donor: the name of the newest donor
+    :param newest_donation: the amount the new donor donated
+    """
+    thank_you_message = "Thank you {} for you generous donation " \
+                        "of ${:.2f}.".format(newest_donor.title(), round(newest_donation,2))
+    print(thank_you_message)
+    print()
+
+def reportHeader():
+    """
+    The header of the report
+    """
+    print("{:25s} | {:11s} | {:9s} | {:12s}".format(
+        "Donor Name", "Total Given", "Num Gifts", "Average Gift"))
+    print("-" * 66)
+
+def createReport(donor_list):
+    """
+    Creates a table of the report that shows the donor names, how many times they donated
+    the average donation and total donations
+    :param donor_list: the list of donors and their donation amounts
+    """
+    total_give = 0.00
+    num_gifts = 0
+    avg_gift = 0
+    report_dict = {}
+    for donor, donations in donor_list.items():
+        num_gifts = len(donations)
+        total_give = sum(donations)
+        avg_gift = total_give / num_gifts
+        report_dict[donor] = [round(total_give,2), num_gifts, round(avg_gift, 2)]
+    reportHeader()
+    for donor, metrics in report_dict.items():
+        print("{:25s}   {:11.2f}   {:9d}   {:12.2f}".format(donor, metrics[0], metrics[1], metrics[2]))
+
+def writeToFile(donor_list):
+    """
+    Writes thank you letters for each individual that has donated and saves the file
+    in the current working directory.
+    :param donor_list: the dictionary of donors and the values each donor has donated
+    """
+    letter = None
+    filename = None
+    donor_file_name = None
+    complete_file_name = None
+    #path = 'C:\\Users\\navgill\Wi2018-Classroom\students\\navdeep\session04\Letters'
+    for donor,donations in donor_list.items():
+        donor_file_name = donor.replace(" ", "_")
+        full_file_name = "{}.txt".format(donor_file_name)
+        complete_file_name = os.path.join(os.getcwd(), full_file_name)
+        letter = letterToSend(donor,donations)
+        with open(complete_file_name,'w+') as f:
+            f.write(letter)
+        f.close()
```

Comment:
When you use a with statement, as you did here, you don't actually need to close the file handle.